### PR TITLE
pass Buffer directly to Rust

### PR DESCRIPTION
**Pull request by sokra**
avoids converting Buffer to string to Buffer to rust string

improves performance

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| buildDuration | 25.2s | 24.9s | -368ms |
| buildDurationCached | 5s | 4.9s | -109ms |
| nodeModulesSize | 332 MB | 332 MB | ⚠️ +237 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.183 | 4.197 | ⚠️ +0.01 |
| / avg req/sec | 597.68 | 595.62 | ⚠️ -2.06 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.18 | 2.17 | -0.01 |
| /error-in-render avg req/sec | 1146.77 | 1152 | +5.23 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 71.9 kB | 71.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.23 kB | 1.23 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| index.html gzip | 523 B | 523 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| buildDuration | 27s | 27s | -44ms |
| buildDurationCached | 4.8s | 4.9s | ⚠️ +57ms |
| nodeModulesSize | 332 MB | 332 MB | ⚠️ +237 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.235 | 4.356 | ⚠️ +0.12 |
| / avg req/sec | 590.35 | 573.94 | ⚠️ -16.41 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.204 | 2.31 | ⚠️ +0.11 |
| /error-in-render avg req/sec | 1134.22 | 1082.31 | ⚠️ -51.91 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.22 kB | 1.22 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: 74639dbd7bc1200241dacc9d3659e94d853e9d17

File: packages/next/build/webpack/loaders/next-swc-loader.js [hunk 1]
```diff
@@ -191,3 +191,6 @@ export default function swcLoader(inputSource, inputSourceMap) {
       }
     )
 }
+
+// accept Buffers instead of strings
+export const raw = true
```
Comment (kdy1) [inline packages/next/build/webpack/loaders/next-swc-loader.js:196]: What is this?
It seems like it's not used by anywhere.

File: packages/next/build/swc/src/transform.rs [hunk 2]
```diff
@@ -117,7 +117,27 @@ where
...
-    let src = cx.get::<JsString>(0)?.into_utf8()?.as_str()?.to_owned();
+    let unknown_src = cx.get::<JsUnknown>(0)?;
+    let src = match unknown_src.get_type()? {
+        napi::ValueType::String => napi::Result::Ok(
+            unsafe { unknown_src.cast::<JsString>() }
+                .into_utf8()?
+                .as_str()?
+                .to_owned(),
+        ),
+        napi::ValueType::Object => napi::Result::Ok(
```
Comment (kdy1) [inline packages/next/build/swc/src/transform.rs:128]: This is undefined behavior. 
https://docs.rs/napi/1.7.7/napi/struct.JsUnknown.html#method.cast

Says
>  V must be match with the return value of get_type

but the returned value is `Object`.

Comment (kdy1) [reply packages/next/build/swc/src/transform.rs:128]: Also, `unsafe` is not required.


According to the doc, `JsBuffer` implements `TryFrom<JsUnknown>`.
https://docs.rs/napi/1.7.7/napi/struct.JsBuffer.html#impl-TryFrom%3CJsUnknown%3E

```rust
JsBuffer::try_from(unknown_src)?
```

will work.

File: packages/next/build/swc/src/transform.rs [hunk 3]
```diff
@@ -117,7 +117,27 @@ where
 {
     let c = get_compiler(&cx);
 
-    let src = cx.get::<JsString>(0)?.into_utf8()?.as_str()?.to_owned();
+    let unknown_src = cx.get::<JsUnknown>(0)?;
+    let src = match unknown_src.get_type()? {
+        napi::ValueType::String => napi::Result::Ok(
+            unsafe { unknown_src.cast::<JsString>() }
```
Comment (kdy1) [inline packages/next/build/swc/src/transform.rs:123]: Same here. We don't need unsafe as 

```rust
JsString::try_from(unknown_src)?
```

will work.

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: packages/next/build/webpack/loaders/next-swc-loader.js (hunk 1 above)
Comment (sokra) [reply packages/next/build/webpack/loaders/next-swc-loader.js:196]: https://webpack.js.org/api/loaders/#raw-loader

File: packages/next/build/swc/src/transform.rs (hunk 3 above)
Comment (sokra) [reply packages/next/build/swc/src/transform.rs]: yep, I changed it

**Review by kdy1 — APPROVED** [review]
Cool!
It's all fine now.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| buildDuration | 17.8s | 17.9s | ⚠️ +83ms |
| buildDurationCached | 3.7s | 3.6s | -95ms |
| nodeModulesSize | 332 MB | 332 MB | ⚠️ +250 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.79 | 2.838 | ⚠️ +0.05 |
| / avg req/sec | 896.01 | 880.99 | ⚠️ -15.02 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.432 | 1.394 | -0.04 |
| /error-in-render avg req/sec | 1745.35 | 1793.54 | +48.19 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 71.9 kB | 71.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.23 kB | 1.23 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| buildDuration | 19.6s | 19s | -600ms |
| buildDurationCached | 3.7s | 3.7s | ⚠️ +11ms |
| nodeModulesSize | 332 MB | 332 MB | ⚠️ +250 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.882 | 2.89 | ⚠️ +0.01 |
| / avg req/sec | 867.44 | 864.97 | ⚠️ -2.47 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.389 | 1.391 | 0 |
| /error-in-render avg req/sec | 1799.83 | 1797.1 | ⚠️ -2.73 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.22 kB | 1.22 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js perf/pass-buffer-to-swc | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: 8c9da9b71f9f574f7b2ed60683713bcaab00256e

**Pull request merged**